doc/topics/async-system.md

```diff
@@ -0,0 +1,209 @@
+# AsyncSystem {#async-system}
+
+[AsyncSystem](@ref CesiumAsync::AsyncSystem) is Cesium Native's primary mechanism for managing asynchronous, multi-threaded work. Understanding it is essential for using, modifying, and extending Cesium Native.
+
+An example will help to illustrate why we need `AsyncSystem` at all.
+
+@mermaid{async-system-load-example}
+
+Let's say we're trying to load a tile in a [3D Tiles](https://github.com/CesiumGS/3d-tiles) tileset. First, we need to start a network request to download the tile's content. We'll do an HTTP GET for a particular URL, and some time tens to hundreds to even thousands of milliseconds later, the tile content at that URL will be downloaded and ready for us to use.
+
+Once it's downloaded, we need to process that tile content that we just downloaded. That means:
+
+* Parse the JSON it contains,
+* Decompress the Draco-encoded meshes or the JPEG-encoded images,
+* Process the triangle meshes for use with a physics engine,
+* And lots more!
+
+We want to do as much of this kind of work as we possibly can in a background thread, because CPUs have multiple cores these days and the main thread is usually busy doing other important work.
+
+Unfortunately, there is inevitably at least a little bit of tile loading work that can't be done in a background thread, though. Game engines, for example, usually have strict rules against creating game objects anywhere other than the main thread. We can't get anything onto the screen without creating game objects, so after the background work is complete, we then need to continue processing this tile in the main thread in order to do the final preparation to render it.
+
+`AsyncSystem` gives us an elegant way to express this kind of sequential process involving a series of asynchronous steps.
+
+## Creating an AsyncSystem {#creating-an-asyncsystem}
+
+Most applications have a single `AsyncSystem` that is used throughout. It is constructed from an [ITaskProcessor](@ref CesiumAsync::ITaskProcessor) instance, which is a simple interface used to perform some work in a background thread. The simplest possible implementation looks like this:
+
+\snippet{trimleft} ExamplesAsyncSystem.cpp simplest-task-processor
+
+This implementation will work well, but it isn't very efficient because a brand new thread is created for each task. Most applications will implement this interface using a thread pool, task graph, or similar functionality that their application already contains.
+
+The `AsyncSystem` can then be created as follows:
+
+\snippet{trimleft} ExamplesAsyncSystem.cpp create-async-system
+
+`AsyncSystem` instances can be safely and efficiently stored and copied by value. This makes it easy to make them available wherever they're needed, including in lambda captures. You can think of an `AsyncSystem` instance as acting a bit like a smart pointer.
```

Comment:
I don't think "facade" is an accurate description of what's happening here. It's not like AsyncSystem is providing an alternate/simplified interface to a complex system. It's that an _instance_ of `AsyncSystem` conceptually acts like a _reference_. If I do this:

```cpp
AsyncSystem asyncSystem = getAsyncSystem();
AsyncSystem copy = asyncSystem;
```

I don't have two independent async systems now. I have one system and two references to it. It's exactly the same as doing this:

```cpp
std::shared_ptr<IAssetAccessor> pAssetAccessor = getAssetAccessor();
std::shared_ptr<IAssetAccessor> pCopy = pAssetAccessor;
```

And that's why I tried to describe it as being smart-pointer-like.

Probably the difficulty in explaining this comes from the fact that it's a bad design to begin with. We should probably just pass AsyncSystem around with shared_ptr just like we do with the other things in TilesetExternals. It would be a bit more verbose, but easier to explain. And less in need of explanation because it would behave the way people expect.

I'm not sure I can justify making that change now, though. So I'll just try to make the explanation here clearer.

Reply:
Why is the distinction important to contributors to Cesium Native, as opposed to "integrators" or implementors of ports? Even calling it a "smart reference" seems like to much detail at this level, and the detail that the underlying implementation object doesn't get destroyed until the last reference is gone belongs in an implementor's guide, not here IMHO.

Reply:
Sorry I'm not quite sure what you're getting at here Tim. IMO this doc is useful for both integrators (people "using" cesium-native) and for contributors (people "developing" cesium-native). It even says so in the second sentence of this doc. 😄 

I would also assert that the copy semantics of the AsyncSystem class are essential to understand for both audiences. Passing around AsyncSystems comes up a _lot_ both within cesium-native and when using it. If you don't understand the copy semantics, you might think it's ok to, for example, capture it by reference, when that is dangerous and mostly unnecessary. Or you might think you need to wrap it up in a `std::make_shared` or something like that, when there's no need at all.

It's true that the note about "the underlying implementation object doesn't get destroyed until the last reference is gone" is mostly only important for people integrating cesium-native into their own applications. Well... except maybe it comes up a bit when writing tests within cesium-native, too. But in any case, this doc doesn't need to be _exclusively_ for contributors even if contributors are our focus right now, right?

Reply:
I'll withdraw my concerns about "Contributors" vs "Implementors"; you're totally correct that this document should be useful to both. But I think this section could still be clearer, and while I think there's a place in this document for a sample implementation, putting it right in the introduction to AsyncSystem creates more questions than perhaps it answers:
- Why copy `AsyncSystem` objects around when they can just be created? Well, Cesium Native can't create one; it's passed in on startup more-or-less and then passed around.
- Why does the creation function return a reference if `AsyncSystem` has copy semantics?

Instead of sending endless criticisms your way, I think I need to actually do my Editor job and write what I think would  flow better here.